File: 백준/Silver/26170. 사과 빨리 먹기/사과 빨리 먹기.java
import java.util.*;
import java.io.*;

class Main{
    final static int INF = 987654321;
    static int [][] map = new int[6][6]; // 보드
    static int [][] visit = new int[6][6]; // 방문여부
    static int R; //현재 위치 x좌표
    static int C; //현재 위치 y좌표
    
    static int dx[] = {-1,1,0,0}; // 좌우 이동
    static int dy[] = {0,0,-1,1}; // 상하 이동
    static int cnt = INF; // 이동횟수

     public static void main(String[] args)throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        for(int i = 1; i<6; i++){
            StringTokenizer st = new StringTokenizer(br.readLine());
            for(int j = 1; j<6; j++){
                map[i][j] = Integer.parseInt(st.nextToken()); // 사과 위치 입력
            }
        }

        StringTokenizer st = new StringTokenizer(br.readLine());
        R = Integer.parseInt(st.nextToken());
        C = Integer.parseInt(st.nextToken());

        visit[R+1][C+1] = 1; // 1씩 더해주어 배열위치 맞춰줌
        
        dfs(R+1,C+1,0); // 탐색 시작

        if(cnt == INF) // 현재 위치에서 사과를 찾을 수 없을 경우
            System.out.println(-1);
        else
            System.out.println(cnt - 1);
    }
    static void dfs(int x, int y, int apple){
        if(apple == 3){ // 사과 3개 발견시,
            cnt = Math.min(cnt, visit[x][y]); // 최소 이동거리 cnt에 저장
            return;
        }

        for(int i =0; i<4; i++){
            int nx = x + dx[i]; 
            int ny = y + dy[i]; // 상하좌우 이동

            if(nx<=0 || ny<=0 || nx >= 6|| ny>= 6) // 범위를 벗어나는 값일 경우
                continue;   // 다시 반복
            if(visit[nx][ny] != 0) // 방문한적 있으면
                continue;   // 다시 반복
            if(map[nx][ny] == -1) // 장애물 존재시
                continue;         // 다시 반복
            
            visit[nx][ny] = visit[x][y] + 1; // 이동횟수 1 추가
            if(map[nx][ny] == 1) // 사과 발견 시
                dfs(nx,ny,apple+1); // 사과 추가해주고 계속 탐색
            else                    // 사과 미발견 시
                dfs(nx,ny,apple); // 계속 탐색진행
            visit[nx][ny] = 0; // 경로 초기화 
        }
    }
}

File: 백준/Silver/26170. 사과 빨리 먹기/사과 빨리 먹기.py
import sys

def dfs(road, x, y,count,apple,answer):
    if apple == 3:
        answer.append(count)
    dx = [0,-1,0,1]
    dy = [-1,0,1,0]
    for i in range(4):	#4개의 방향으로 돈다
        ddx = x + dx[i]
        ddy = y + dy[i]
        if ddx < 0 or ddx > 4 or ddy < 0 or ddy > 4: #범위가 아니라면 pass
            continue
        if road[ddx][ddy] == -1:	#해당 좌표가 -1 이라면 막혔으니 pass
            continue
        if road[ddx][ddy] == 1:	#1이라면 status를 True로 정해주고 apple의 수를 증가한다.
            status = True
            apple+=1
        else:
            status = False	#0일 경우에는 status를 False로 한다. 
        cnt = road[x][y]	#dfs에서 빠져 나왔을 경우 road[x][y]의 값을 복귀시켜야 하기 때문에 변수에 저장시켜 놓는다.
        road[x][y] = -1
        dfs(road, ddx,ddy,count+1, apple,answer)
        road[x][y] = cnt	#빠져나왔을 경우 좌표 값 원복
        if status == True:	#만약 사과를 하나 챙겼다면 원복
            apple -=1
    return answer
road = []
for i in range(5):
    a = list(map(int, sys.stdin.readline().split()))
    road.append(a)
x,y = map(int, sys.stdin.readline().split())
apple = 0
answer = []
if road[x][y] == 1:
    apple +=1
answer = dfs(road,x,y,0,apple, answer)
if len(answer) == 0:
    print(-1)
else:
    print(min(answer))